src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPEntityListener.java

```diff
@@ -49,41 +54,58 @@ public MVNPEntityListener(MultiverseNetherPortals plugin) {
         this.worldManager = this.plugin.getCore().getMVWorldManager();
         this.pt = new PermissionTools(this.plugin.getCore());
         this.playerErrors = new HashMap<String, Date>();
-        this.eventRecord = new HashMap<String, Location>();
         this.messaging = this.plugin.getCore().getMessaging();
         this.locationManipulation = this.plugin.getCore().getLocationManipulation();
-
+        this.eventRecord = new HashMap<>();
+        this.eventRecord.put(PortalType.ENDER, new HashSet<>());
+        this.eventRecord.put(PortalType.NETHER, new HashSet<>());
     }
 
-    protected void shootPlayer(Player p, Block block, PortalType type) {
+    protected boolean shootPlayer(Player p, Block block, PortalType type) {
         if (!plugin.isUsingBounceBack()) {
-            this.plugin.log(Level.FINEST, "You said not to use bounce back so the player is free to walk into portal!");
-            return;
+            this.plugin.log(Level.FINEST, "Bounceback is disabled, so the player is free to walk into portals!");
+            return false;
         }
+
         this.playerErrors.put(p.getName(), new Date());
-        double myconst = 2;
         double newVecX = 0;
         double newVecZ = 0;
-        // Determine portal axis:
-        BlockFace face = p.getLocation().getBlock().getFace(block);
-        if (block.getRelative(BlockFace.EAST).getType() == Material.NETHER_PORTAL || block.getRelative(BlockFace.WEST).getType() == Material.NETHER_PORTAL) {
-            this.plugin.log(Level.FINER, "Found Portal: East/West");
-            if (p.getLocation().getX() < block.getLocation().getX()) {
-                newVecX = -1 * myconst;
+        double strength = 2;
+        boolean playerBounced = false;
+
+        StringBuilder debugMessage = new StringBuilder().append("Player: ").append(p.getName());
+        if (type == PortalType.ENDER) {
+            debugMessage.append(" entered an End Portal. There is currently no bounceback implementation for End Portals.");
+        } else if (type == PortalType.NETHER) {
+            // Determine portal axis:
+            if (block.getRelative(BlockFace.EAST).getType() == Material.NETHER_PORTAL || block.getRelative(BlockFace.WEST).getType() == Material.NETHER_PORTAL) {
+                // we add 0.5 to the location of the block to get the center
+                if (p.getLocation().getZ() < block.getLocation().getZ() + 0.5) {
+                    debugMessage.append(" entered Nether Portal from the North");
+                    newVecZ = -1 * strength;
+                } else {
+                    debugMessage.append(" entered Nether Portal from the South");
+                    newVecZ = 1 * strength;
+                }
```

Comment:
This method could do with some serious refactoring. It's very hard to follow what is going on here. (Not your fault)